transcom/mymove
# Persists payment service items and associated records with payment requests

## Description

This PR persists payment service items, payment service item parameters, and other associated data as part of the payment request creation process.

## Reviewer Notes

There are some tests, but I'm still working on adding a more complete set.  I also need to add more logging and deal with a few other smaller things.  But the persistence should work.

Note that I had to comment out some of the YAML around proof of service documents for now to get past Swagger validations.  We can add those back in with the proof of service document story.

I also commented out some of the model validation for pricing since we're not putting any pricing in yet with these payment service items.

## Setup

1. If this is your first time connecting via Mutual TLS, make sure `primelocal` is set in your `/etc/hosts` file ([reference](https://github.com/transcom/mymove#setup-hosts))
1. `make server_build`
1. `make db_dev_e2e_populate`
1. `make server_run`

### To test with `curl`, paste this command into a terminal window:

```
curl -X POST \
  -k \
  --cert ./config/tls/devlocal-mtls.cer  \
  --key ./config/tls/devlocal-mtls.key \
  https://primelocal:9443/prime/v1/payment-requests \
  -H 'Content-Type: application/json' \
  -d '{"isFinal": false, "moveTaskOrderID": "5d4b25bb-eb04-4c03-9a81-ee0398cb779e", "serviceItems": [{"id": "9db1bf43-0964-44ff-8384-3297951f6781", "params": [{"key": "weight", "value": "1234"}, {"key": "pickup", "value": "2019-12-16"}]}, {"id": "d886431c-c357-46b7-a084-a0c85dd496d3", "params": [{"key": "weight", "value": "5678"}]}]}'
```

### To test with Postman:

1. Make sure your client cert is set up: Go to the wrench icon, then `Settings`, then `Certificates`.  Set the CRT file and KEY file to match the paths in the `curl` command above.  Use `primelocal:9443` for the host and port.
1. In `Settings`, under `General`, turn off `SSL certificate verification`.
1. Create a new `POST` request to `https://primelocal:9443/prime/v1/payment-requests`
1. In the `Body` tab, click the `raw` radio button, and paste in this JSON: `{"isFinal": false, "moveTaskOrderID": "5d4b25bb-eb04-4c03-9a81-ee0398cb779e", "serviceItems": [{"id": "9db1bf43-0964-44ff-8384-3297951f6781", "params": [{"key": "weight", "value": "1234"}, {"key": "pickup", "value": "2019-12-16"}]}, {"id": "d886431c-c357-46b7-a084-a0c85dd496d3", "params": [{"key": "weight", "value": "5678"}]}]}`
1. Change the drop-down to the right of the radio buttons to `JSON` (the header should then automatically set `Content-Type` to `application/json`
1. Hit the `Send` button

### OLD swagger way (this won't work anymore now that I've merged in mutual TLS code):

To test out the `payment-requests` endpoint in the Swagger UI:
1. `make clean server_build client_build`
1. `make db_dev_reset db_dev_e2e_populate`
1. `make server_run`
1. `make client_run`
1. Log into http://officelocal:3000/
1. Go to the Swagger UI: http://officelocal:3000/prime/v1/docs#/
1. Try out the `POST /payment_requests`

You can use this payload and then hit `Execute`:

```
{
  "isFinal": false,
  "moveTaskOrderID": "5d4b25bb-eb04-4c03-9a81-ee0398cb779e",
  "serviceItems": [
    {
      "id": "9db1bf43-0964-44ff-8384-3297951f6781",
      "params": [
        {
          "key": "weight",
          "value": "1234"
        },
        {
          "key": "pickup",
          "value": "2019-12-16"
        }
      ]
    },
    {
      "id": "d886431c-c357-46b7-a084-a0c85dd496d3",
      "params": [
        {
          "key": "weight",
          "value": "5678"
        }
      ]
    }
  ]
}
```

## Code Review Verification Steps

* Any new migrations/schema changes:

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-622) for this change
